Navigation Area block

adamziel opened the pull request:
## Description

This PR [proposes preserving navigation on theme switch](https://github.com/WordPress/gutenberg/issues/36087) by introducing a "Navigation Area" block.

**Relies on https://github.com/WordPress/gutenberg/pull/36212**

The gist here is that:
* themes define areas
* if theme A's areas match theme B's areas we then reuse the navigation CPT
* theme A and B in their templates use a navigation area block to wrap their menus
* menu to area relationship is saved via site options like `{ "primary": 794, "secondary": 112 }`

Theme authors would ship the following HTML in their templates:

```html
<!-- wp:navigation-area {"area":"primary"} -->
<!-- wp:navigation /-->
<!-- /wp:navigation-area -->
```

Which makes the nested navigation block use the `wp_navigation` post associated with the `primary` area.

Changing saving the navigation area changes the post ID to which the primary area is mapped. The nice part is that this PR uses the entity system, which gives us things like in-editor synchronization and confirmation dialogs for free.

See this a short video from the site editor. I wrestled with GB a bit but it worked more-or-less as intended, maybe except of the unsaved menu:

https://user-images.githubusercontent.com/205419/140085876-85fe37e0-afc4-43cf-a6f8-0713d5b9b1d8.mp4

Switching to another theme just works.

## About design choices

* I used site_options because this is a minimal mapping that doesn't need anything fancier. Taxonomies could go wrong in that they can express multiple posts pointing to the same area. This mapping can't.
* The navigation block understands the concept of an area so that it can call `setAreaMenu` if another menu is selected.

What's missing:
* Linting
* Unit tests for the endpoint
* Group block-like selection mechanics for the Navigation Area block

adamziel committed:
try nav-area-block

adamziel committed:
Resolve conflicts with trunk

adamziel commented:
Something like this could be useful in the `navigation/index.php`, too:

```php
	if ( ! empty( $block->context['navigationArea'] ) ) {
		$area = $block->context['navigationArea'];
		$mapping = get_option( 'fse_navigation_areas', array() );
		if ( ! empty( $mapping[ $area ] ) ) {
			$attributes[ 'navigationMenuId' ] = $mapping[ $area ];
		}
	}
```

Edit: I just added it

adamziel committed:
Rely on contexts also in PHP navigation block's render callback

talldan committed:
Lint and typo fixes

talldan committed:
Add a description for the navigation area block

talldan committed:
Remove navigation area and unused functions

talldan commented on packages/block-library/src/navigation-area/edit.js:
```diff
@@ -0,0 +1,94 @@
+		orientation: 'horizontal',
+		renderAppender: false,
+		template: TEMPLATE,
+		// templateLock: "insert",
```
Just wondering why this was commented out. Was there an issue with it?

I was thinking we could try `templateLock: 'all'`. There are also some new block level options introduced here to prevent removal of the nav block - https://github.com/WordPress/gutenberg/pull/32457.

talldan committed:
Update navigation area example

talldan committed:
Remove unused attribute

talldan committed:
Minor copy update

talldan committed:
Try passing the correct menu id into the template

talldan committed:
Only build template when menu id is known and add loading state

talldan committed:
Show navigation block placeholder when in an unassigned navigation area

talldan committed:
Support deleting menus in navigation areas

talldan commented:
In testing, it's working very well. Switching themes is working flawlessly when they have these navigation areas defined. I don't see any reason not to progress with this. It'd be a matter of priority to update bundled theme templates with this new block and to get the word out to theme developers about it (I'll add 'Needs dev note').

There's a small bit of a glitch when switching menus sometimes—on the odd occasion it considers the menu 'unsaved'. I think this is a bug in `trunk` though, so I'll work on fixing that as a separate effort.

Some potential opportunities:
- We can preload the navigation menus for each defined navigation area. This is great as they can take a little while to load.
- We could look into migrating `__unstableLocation` to this block for universal themes. (not needed for 5.9)

talldan committed:
Add location keyword

talldan committed:
Remove unused stylesheet

draganescu commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,331 @@
+	}
+
+	/**
+	 * Retrieves all menu locations, depending on user context.
```
Let's not use "locations" as it confuses things with classic locations. It'a areas.

draganescu commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,331 @@
+	}
+
+	/**
+	 * Checks if a given request has access to read a menu location.
```
Same location -> area

draganescu commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,331 @@
+	 * @return bool|WP_Error True if the request has access to update the item, false or WP_Error object otherwise.
+	 */
+	public function update_item_permissions_check( $request ) {
+		// @TODO
```
what is the todo here?

draganescu commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,331 @@
+	}
+
+	/**
+	 * Retrieves a specific menu location.
```
location -> area

draganescu commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,331 @@
+		 * @param object $area The original status object.
+		 * @param WP_REST_Request $request Request used to generate the response.
+		 */
+		return apply_filters( 'rest_prepare_menu_location', $response, $area, $request );
```
wouldn't we need an "area" specific filter instead?

talldan reviewed (approved):
Tentatively giving this approval, so that there's no issue with it being merged in time for 5.9.

Really happy with how it works, and it looks like there are only minor changes and tests remaining. Look forward to getting this shipped, great work @adamziel!

adamziel committed:
Replace references to menu locations with navigation areas

adamziel committed:
Remove links from the first version

adamziel committed:
Add gutenberg_register_navigation_areas that enables overriding the default navigation areas.

adamziel committed:
Rename "menu" property returned by the REST API to "navigation"

adamziel committed:
Replace a call to get_item() in update_item() with an encapsulation (common get_navigation_area_object function)

adamziel committed:
Remove @Validate comment and follow the menu-items endpoint in not validating the db object ID.

It would be great to add one at some point, but right now the way
gutenberg saves data is by issuing multiple concurrent http requests, so
there is a huge potential for race conditions.

adamziel committed:
Remove @TODOs related to changing permissions checks. The checks are fine.

adamziel committed:
Restore the // Experimental blocks. comment

adamziel committed:
Preload navigation areas in the site editor

adamziel committed:
Use templateLock: 'all' for navigation area inner blocks

adamziel committed:
Add missing fixtures

adamziel committed:
Add file doc comment

adamziel committed:
Remove navigation menu bugfix so that it can be extracted into another PR

adamziel committed:
Adjust fixtures

adamziel committed:
Adjust fixtures

adamziel committed:
Adjust fixturs

adamziel commented:
Thank you so much for pushing this PR forward so much @talldan! I polished all the kinks we discussed and I think we may be good to go now! Would you please take a look and confirm?

Important! I moved the core-data bugfix (which this PR relies on) to https://github.com/WordPress/gutenberg/pull/36212. ~~Please apply #36212 before testing this one.~~ It's already merged.

I also found a bug in the navigation block and proposed a fix in https://github.com/WordPress/gutenberg/pull/36210.

anton-vlasenko commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,284 @@
+	 */
+	public function get_item_schema() {
+		$schema = array(
+			'$schema'    => 'http://json-schema.org/draft-04/schema#',
```
Is it a typo?
```suggestion
			'schema'    => 'http://json-schema.org/draft-04/schema#',
```

draganescu replied:
I think so. Suggestion commited.

talldan replied:
Every definition of `get_item_schema` in WordPress core has `'$schema'`, so I'm going to revert this.

anton-vlasenko replied:
It looked weird when I saw it. I should have checked the usage of it. Sorry for the confusion.

draganescu committed:
Update lib/class-wp-rest-block-navigation-areas-controller.php

Co-authored-by: Anton Vlasenko <43744263+anton-vlasenko@users.noreply.github.com>

draganescu commented on lib/rest-api.php:
```diff
@@ -40,6 +40,15 @@ function gutenberg_register_rest_menu_location() {
+ * Registers the navigation areas REST API routes.
+ */
+function gutenberg_register_rest_navigation_areas() {
+	$nav_menu_location = new WP_REST_Block_Navigation_Areas_Controller();
```
$nav_menu_location -> $nav_menu_area

talldan replied:
I fixed it. Looks like it was copy/pasted from the line above.

Also fixed customizer_nonces below which also copy/pasted!

draganescu commented on packages/block-library/src/navigation-area/edit.js:
```diff
@@ -0,0 +1,105 @@
+					<SelectControl
+						label={ _x( 'Area' ) }
+						value={ area }
+						// `undefined` is required for the preload attribute to be unset.
```
what is this comment for?

adamziel replied:
Dev artifact, I copied that code from somewhere, lets remove it

draganescu reviewed (approved):
I think this is a very solid foundation. 

- `gutenberg_register_navigation_areas` should be filtered so that default areas can be extended
- there is still one place where it says "location" instead of "area"
- storing area to menu CPT association in `options` sounds better to me the more I thought about it ...

talldan commented on packages/block-library/src/navigation-area/save.js:
```diff
@@ -0,0 +1,12 @@
+
+export default function save() {
+	return (
+		<div { ...useBlockProps.save() }>
```
I think we can remove the div and only return `InnerBlocks.Content` link the nav block does.

talldan replied:
I went ahead and did that in af6136d

talldan committed:
Remove div from save content

talldan committed:
Fix REST API registration variable nameing

talldan committed:
Revert "Update lib/class-wp-rest-block-navigation-areas-controller.php"

This reverts commit e31fdfbe0f39e2051258c7bd05de45eab3bc85bb.

talldan committed:
Avoid trying to preload post with 0 id, which is default for a navigation area

talldan reviewed (approved):
Approving again for good luck.

adamziel commented:
Merging for good luck.

MaggieCabrera commented:
> We could look into migrating __unstableLocation to this block for universal themes. (not needed for 5.9)

Yes please! Not just for Universal themes but for users who switch from a classic theme to a block theme too

adamziel commented on packages/block-library/src/navigation-area/edit.js:
```diff
@@ -0,0 +1,105 @@
+			navigationAreas?.map( ( { name, description } ) => ( {
+				label: description,
+				value: name,
+			} ) ),
```
Let's do `|| [] ` to prevent a null value making it to MenuItemsChoice

adamziel commented on packages/block-library/src/navigation-area/edit.js:
```diff
@@ -0,0 +1,105 @@
+			);
+			return {
+				navigationAreas: getEntityRecords( 'root', 'navigationArea' ),
+				hasResolvedNavigationAreas: hasFinishedResolution(
```
We could do `&& navigationAreas.length` to prevent pretending everything is fine when there are no areas

talldan commented:
> Yes please! Not just for Universal themes but for users who switch from a classic theme to a block theme too

👍 Absolutely, that's a very good point, we need to get our thinking caps on about the classic theme to block theme migration.

It shouldn't be too difficult, we need hook into theme switch, figure out which classic menu is assigned to a corresponding navigation area, create a wp_navigation post with the menu items converted to blocks, and then assign its id to the navigation area.

getdave commented:
> It shouldn't be too difficult, we need hook into theme switch, figure out which classic menu is assigned to a corresponding navigation area, create a wp_navigation post with the menu items converted to blocks, and then assign its id to the navigation area.

I detailed the existing classic mechanic in https://github.com/WordPress/gutenberg/issues/35750#issuecomment-950755154. Hopefully this serves as inspiration for an automated mapping solution.

I will say again that I think that on Theme switch, the classic menu should be _copied/cloned_ and converted into a block based menu leaving the original in place. If the user switches back to a classic theme, then we should allow that classic theme to simply pull through the original classic menu.

adamziel commented:
Maybe the first, `__unstable` version could just reuse the navigation editor code to bring in the classic menu as blocks on the editor load, and then, on save, store them as a wp_navigation post?

adamziel commented:
Linking this here: [Migrate classic menus to block-based menus on theme switch](https://github.com/WordPress/gutenberg/pull/36255)

spacedmonkey commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,284 @@
+					'callback'            => array( $this, 'update_item' ),
+					'permission_callback' => array( $this, 'update_item_permissions_check' ),
+					'args'                => $this->get_endpoint_args_for_item_schema( WP_REST_Server::EDITABLE ),
+				),
```
Are we going to add support for batching?

spacedmonkey commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,284 @@
+
+		register_rest_route(
+			$this->namespace,
+			'/' . $this->rest_base . '/(?P<area>[\w-]+)',
```
Why query by name and not ID?

spacedmonkey commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,284 @@
+		foreach ( gutenberg_get_navigation_areas() as $name => $description ) {
+			$area          = $this->get_navigation_area_object( $name );
+			$area          = $this->prepare_item_for_response( $area, $request );
+			$data[ $name ] = $this->prepare_response_for_collection( $area );
```
@TimothyBJacobs what do you think if this? Other endpoints simple are a flat array. A key based array is different from other endpoints.

TimothyBJacobs replied:
We do have a couple of object responses, but an array is preferable because the REST API does assume that collection responses are lists.

spacedmonkey commented on lib/class-wp-rest-block-navigation-areas-controller.php:
```diff
@@ -0,0 +1,284 @@
+	private function get_navigation_area_object( $name ) {
+		$available_areas   = gutenberg_get_navigation_areas();
+		$mapping           = get_option( 'fse_navigation_areas', array() );
+		$area              = new stdClass();
```
Could we create a custom object for this?

spacedmonkey reviewed (commented):
Posted some early thoughts. Will review again later.

spacedmonkey commented on lib/navigation.php:
```diff
@@ -517,3 +517,57 @@ function gutenberg_rename_navigation_menus_admin_menu_entry() {
+	);
+	foreach ( $active_areas as $post_id ) {
+		if ( 0 !== $post_id ) {
+			$paths[] = "/wp/v2/navigation/$post_id?context=edit";
```
```
$paths[] = add_query_args( 'context', 'edit', rest_get_route_for_post( $post_id ) );
```

spacedmonkey commented:
Sorry I missed this PR. I added my thoughts on the PR anyway, in hope that some of my questions are answered

talldan commented on lib/navigation.php:
```diff
@@ -517,3 +517,57 @@ function gutenberg_rename_navigation_menus_admin_menu_entry() {
+ *
+ * @param array $new_areas Supported navigation areas.
+ */
+function gutenberg_register_navigation_areas( $new_areas ) {
+	global $gutenberg_navigation_areas;
+	$gutenberg_navigation_areas = $new_areas;
+}
+
+// Register the default navigation areas.
+gutenberg_register_navigation_areas(
+	array(
+		'primary'   => 'Primary',
+		'secondary' => 'Secondary',
+		'tertiary'  => 'Tertiary',
+	)
+);
```
Something I just noticed. If a theme dev is able to change the keys here by calling `register_navigation_areas` it'll degrade the capability of theme migration. This is a problem we have with the existing menus system in WordPress.

An example - one theme calls a menu location 'Header', the other 'Primary', and when switching no migration happens because the names don't match.

The way around this, I think, is to keep an internal representation of the area `primary`, `secondary` (or maybe even just an index `0`, `1`), but allow users to set a name that's used in the UI ('Header', 'Footer').

I quite like the idea of:
```php
	array(
		'primary'   => 'Header',
		'secondary' => 'Sidebar',
		'tertiary'  => 'Footer',
	)
```

But then what if I have four or five or ten navigation areas? What comes after tertiary?

It's a difficult problem, but I think we should try to solve it for 5.9 by improving this function. Lets come up with some ideas for this.

getdave replied:
I think we should provide some strong defaults but [allow users to change extend how they like](https://github.com/WordPress/gutenberg/issues/36373).

Most devs will stick with the convention that comes by default so that will solve 80% use case of Theme switch.

If a Theme dev choose to alter the area name then they will need to handle the situation where navigations don't migrate across on Theme switch. This provides a big incentive to stick with the registered defaults.

I didn't realise that `register_navigation_areas` would overwrite the existing items. If that's the case then I think we should make that function private and instead provide a filter. This approach is more "WordPress" and also tends to encourage extension rather than wholesale overwriting.


> What comes after tertiary?

`quaternary, quinary, senary, septenary, octonary, nonary, denary.`

talldan replied:
I've added a comment over here - https://github.com/WordPress/gutenberg/issues/36373#issuecomment-968511319
